std/algorithm/iteration.d

```diff
@@ -4280,3 +4284,218 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Finds the mean (colloquially known as the average) of a range. This function
+will return $(D real.nan) if the range is empty. This function is
+$(BIGOH r.length).
+
+Params:
+    r = an input range
+
+Returns:
+    the mean of r
+*/
+real mean(R)(R r)
+if (isInputRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : real))
+{
+    if (r.empty)
+    {
+        return real.nan;
+    }
+
+    static if (hasLength!R)
+    {
+        if (r.length == 1)
+        {
+            return r.front;
+        }
+
+        return sum(r) / real(r.length);
+    }
+    else
+    {
+        import std.typecons : tuple;
+
+        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
```

Comment:
It's an issue of reductive algorithms in general. The information about how many iterations were performed is lost, but it scales poorly to modify each generic algorithm to optionally retain the number of iterations.

Maybe the solution is a higher order range that wraps an input range and tracks the number of pops. After being passed to a reductive algorithm it would result in presenting the effective length of the underlying range. I'm not sure how it should interact with reference-ness and `save`.

In regards to this PR, maybe it's appropriate to leave a comment explicitly stating the accuracy difference between the two paths.